plugins/inputs/slurm/slurm.go

```diff
@@ -0,0 +1,666 @@
+		int32Ptr *int32
+		strPtr   *string
+		ok       bool
+	)
+
+	for _, partition := range partitions {
+		records := make(map[string]interface{})
+		tags := make(map[string]string)
+
+		tags["source"] = s.baseURL.Hostname()
+		strPtr, ok = partition.GetNameOk()
+		if !ok {
+			continue
+		}
+		tags["name"] = *strPtr
+
+		strPtr, ok = partition.GetStateOk()
+		if !ok {
+			continue
+		}
+		records["state"] = *strPtr
+
+		int32Ptr, ok = partition.GetTotalCpusOk()
+		if !ok {
+			continue
+		}
+		records["total_cpu"] = *int32Ptr
+
+		int32Ptr, ok = partition.GetTotalNodesOk()
+		if !ok {
+			continue
+		}
+		records["total_nodes"] = *int32Ptr
+
+		strPtr, ok = partition.GetNodesOk()
+		if !ok {
+			continue
+		}
+		records["nodes"] = *strPtr
+
+		strPtr, ok = partition.GetTresOk()
+		if !ok {
+			continue
+		}
+		records["tres"] = *strPtr
+
+		acc.AddFields("slurm_partitions", records, tags)
+	}
+}
+
+func (s *Slurm) GatherReservationsMetrics(acc telegraf.Accumulator,
+	reservations []goslurm.V0038Reservation) {
+	var (
+		int32Ptr *int32
+		strPtr   *string
+		ok       bool
+	)
+	for _, reservation := range reservations {
+		records := make(map[string]interface{})
+		tags := make(map[string]string)
+
+		tags["source"] = s.baseURL.Hostname()
+		strPtr, ok = reservation.GetNameOk()
+		if !ok {
+			continue
+		}
+		tags["name"] = *strPtr
+
+		int32Ptr, ok = reservation.GetCoreCountOk()
+		if !ok {
+			continue
+		}
+		records["core_count"] = *int32Ptr
+
+		int32Ptr, ok = reservation.GetCoreSpecCntOk()
+		if !ok {
+			continue
+		}
+		records["core_spec_count"] = *int32Ptr
+
+		strPtr, ok = reservation.GetGroupsOk()
+		if !ok {
+			continue
+		}
+		records["groups"] = *strPtr
+
+		strPtr, ok = reservation.GetUsersOk()
+		if !ok {
+			continue
+		}
+		records["users"] = *strPtr
+
+		int32Ptr, ok = reservation.GetStartTimeOk()
+		if !ok {
+			continue
+		}
+		records["start_time"] = *int32Ptr
+
+		strPtr, ok = reservation.GetPartitionOk()
+		if !ok {
+			continue
+		}
+		records["partition"] = *strPtr
+
+		strPtr, ok = reservation.GetAccountsOk()
+		if !ok {
+			continue
+		}
+		records["accounts"] = *strPtr
+
+		int32Ptr, ok = reservation.GetNodeCountOk()
+		if !ok {
+			continue
+		}
+		records["node_count"] = *int32Ptr
+
+		strPtr, ok = reservation.GetNodeListOk()
+		if !ok {
+			continue
+		}
+		records["node_list"] = *strPtr
+
+		acc.AddFields("slurm_reservations", records, tags)
+	}
+}
+
+func (s *Slurm) Gather(acc telegraf.Accumulator) (err error) {
+	auth := context.WithValue(
+		context.Background(),
+		goslurm.ContextAPIKeys,
+		map[string]goslurm.APIKey{
+			"user":  {Key: s.Username},
+			"token": {Key: s.Token},
+		},
+	)
+
+	if !s.endpointMap["diag"] {
+		diagResp, respRaw, err := s.client.SlurmAPI.SlurmV0038Diag(auth).Execute()
+		if err != nil {
+			return err
+		}
+		defer respRaw.Body.Close()
+		diag, ok := diagResp.GetStatisticsOk()
+		if !ok {
+			return fmt.Errorf("error getting diag: %w", err)
+		}
+		s.GatherDiagMetrics(acc, diag)
```

Comment:
`err` will always be `nil` here? Maybe rewrite to this then:
```suggestion
		if diag, ok := diagResp.GetStatisticsOk(); ok {
			s.GatherDiagMetrics(acc, diag)
		}
```
And use the formatted error message above when executing the API call..

Reply:
That's a great idea! Thanks a ton for bringing it up. I added these changes on 2ab967610d012281d93dcebd1be4ed2b75c03b99